compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5795,26 +5795,25 @@ private void checkObjectFunctionInvocationExpr(BLangInvocation iExpr, BObjectTyp
         checkInvocationParamAndReturnType(iExpr);
     }
 
-    // Here, an action invocation can be either of the following two forms:
+    // Here, an action invocation can be either of the following three forms:
     // - foo->bar();
-    // - start foo.bar(); or start foo->bar()
+    // - start foo.bar(); or start foo->bar(); or start (new Foo()).foo();
     private void checkActionInvocation(BLangInvocation.BLangActionInvocation aInv, BObjectType expType) {
-        BLangValueExpression varRef = (BLangValueExpression) aInv.expr;
 
-        if (((varRef.symbol.tag & SymTag.ENDPOINT) != SymTag.ENDPOINT) && !aInv.async) {
-            dlog.error(aInv.pos, DiagnosticErrorCode.INVALID_ACTION_INVOCATION, varRef.getBType());
+        if (aInv.expr.getKind() == NodeKind.SIMPLE_VARIABLE_REF &&
+                (((((BLangSimpleVarRef) aInv.expr).symbol.tag & SymTag.ENDPOINT) !=
+                        SymTag.ENDPOINT) && !aInv.async)) {
```

Comment:
Shall we extract this condition in to it's own variable or a function? `if (!isEndpointReference && !aInv.async) {`

Reply:
will extract that logic to a separate function. that way it will be more clear